automation/clients/airtable.py

```diff
@@ -37,24 +39,63 @@
 
 # TODO : consider not allowing users change the id field
 class BaseModel(pydantic.BaseModel, abc.ABC):
-    id: str
-    created_at: datetime.datetime
+    id: str = pydantic.Field(allow_mutation=False)
+    created_at: datetime.datetime = pydantic.Field(allow_mutation=False)
+
+    _snapshot = pydantic.PrivateAttr(default=None)
 
     class Config:
         allow_population_by_field_name = True
+        underscore_attrs_are_private = True
+        validate_assignment = True
+
+    def snapshot(self):
+        self._snapshot = self.copy(deep=True)
+
+    def get_modified_fields(self):
+        if self._snapshot is None:
+            raise RuntimeError(
+                "Attempted to get modified fields without having taken a "
+                "snapshot"
+            )
+
+        snapshot_dict = self._snapshot.dict()
+        self_dict = self.dict()
+        all_fields = snapshot_dict.keys() | self_dict.keys()
+
+        modified_fields = set()
+        for field in all_fields:
+            if snapshot_dict.get(field) != self_dict.get(field):
+                modified_fields.add(field)
+
+        return modified_fields
 
     @classmethod
     def from_airtable(cls, raw_dict):
-        return cls(
+        model = cls(
             id=raw_dict["id"],
             created_at=raw_dict["createdTime"],
             **raw_dict["fields"],
         )
-
-    def to_airtable(self):
-        fields = self.dict(by_alias=True, exclude_none=True)
-        del fields["id"]
-        del fields["created_at"]
+        model.snapshot()
+        return model
+
+    def to_airtable(self, modified_only=False):
+        include = None
+        if modified_only:
+            include = self.get_modified_fields()
+
```

Comment:
Somewhere in here we need to serialize the `_meta` field to a JSON encoded plain string inside the overall JSON blob.

We won't notice this problem if `automation_prime` never modifies the `_meta` field, and maybe that's a good rule to have so it doesn't clobber what `firebase-functions` is doing, but if we want that rule we should explicitly exclude that field here.

I'll leave it up to you whether you want to add logic for that or just merge now.

Reply:
I'm pretty tied up atm. So I'll go ahead and merge it now.